File: software/ips_jazzy_ws/src/ips_nodes_cpp/src/clock_sync_node.cpp
// =============================================================================
//  clock_sync_node.cpp
//  Drop-in C++ replacement for Python clock_sync_node.py.
//
//  Wraps SyncEngine in an rclcpp::Node. Subscribes to /uwb/anchor_reports +
//  /uwb/session_events, publishes /uwb/corrected_toa + /uwb/sync_status.
//
//  Performance vs Python version:
//    - No print/cout in hot path (was 264+ calls/sec @ 25Hz blink)
//    - No internal engine lock (single-threaded executor serializes)
//    - process_pending_tags only for the relevant slave (4× less work)
//    - Fixed-size Eigen matrices (no heap alloc per blink)
//    - Glitch protection + rollback (fixes 2 mm bias bug)
// =============================================================================
#include <chrono>
#include <memory>
#include <mutex>
#include <string>

#include <rclcpp/rclcpp.hpp>

#include "ips_msgs/msg/corrected_to_a.hpp"
#include "ips_msgs/msg/session_event.hpp"
#include "ips_msgs/msg/sync_status.hpp"
#include "ips_msgs/msg/uwb_anchor_report.hpp"

#include "ips_nodes_cpp/dw1000_constants.hpp"
#include "ips_nodes_cpp/sync_engine.hpp"

namespace ips {

using std::placeholders::_1;

// QoS profiles (must match Python ips_nodes/common.py)
static rclcpp::QoS qos_sensor_best_effort_deep() {
    rclcpp::QoS q(500);                          // bumped from 200 for 25Hz blink
    q.best_effort().keep_last(500);
    return q;
}
static rclcpp::QoS qos_state_reliable() {
    rclcpp::QoS q(20);
    q.reliable().keep_last(20);
    return q;
}

class ClockSyncNode : public rclcpp::Node {
public:
    ClockSyncNode() : rclcpp::Node("clock_sync") {
        // ---- Parameters ----------------------------------------------
        declare_parameter<double>("status_period_s", 1.0);
        declare_parameter<bool>  ("kf_enabled",      true);

        const double status_period =
            get_parameter("status_period_s").as_double();
        const bool   kf_enabled    =
            get_parameter("kf_enabled").as_bool();

        // ---- Engine setup --------------------------------------------
        SyncEngine::Config cfg;
        cfg.kf_enabled = kf_enabled;
        engine_ = std::make_unique<SyncEngine>(cfg);

        engine_->set_toa_callback(
            [this](const CorrectedToAResult& r) { on_corrected_toa(r); });

        engine_->set_log_callback(
            [this](int sev, const std::string& msg) {
                switch (sev) {
                    case 0: RCLCPP_DEBUG(get_logger(), "%s", msg.c_str()); break;
                    case 1: RCLCPP_INFO (get_logger(), "%s", msg.c_str()); break;
                    case 2: RCLCPP_WARN (get_logger(), "%s", msg.c_str()); break;
                    default: RCLCPP_ERROR(get_logger(), "%s", msg.c_str());
                }
            });

        // ---- Publishers ----------------------------------------------
        pub_toa_ = create_publisher<ips_msgs::msg::CorrectedToA>(
            "/uwb/corrected_toa", qos_state_reliable());
        pub_status_ = create_publisher<ips_msgs::msg::SyncStatus>(
            "/uwb/sync_status", qos_state_reliable());

        // ---- Subscriptions -------------------------------------------
        sub_reports_ = create_subscription<ips_msgs::msg::UwbAnchorReport>(
            "/uwb/anchor_reports", qos_sensor_best_effort_deep(),
            std::bind(&ClockSyncNode::on_report, this, _1));

        sub_events_ = create_subscription<ips_msgs::msg::SessionEvent>(
            "/uwb/session_events", qos_state_reliable(),
            std::bind(&ClockSyncNode::on_event, this, _1));

        // ---- Status timer --------------------------------------------
        status_timer_ = create_wall_timer(
            std::chrono::milliseconds(
                static_cast<int>(status_period * 1000)),
            std::bind(&ClockSyncNode::publish_status, this));

        RCLCPP_INFO(get_logger(),
            "clock_sync (C++) ready (KF=%s)",
            kf_enabled ? "on" : "off");
    }

private:
    // ---- Report dispatch ---------------------------------------------
    void on_report(const ips_msgs::msg::UwbAnchorReport::SharedPtr msg) {
        std::lock_guard<std::mutex> lock(engine_mu_);
        try {
            switch (msg->report_type) {
                case ips_msgs::msg::UwbAnchorReport::TYPE_MASTER_CLOCK:
                    engine_->add_master_clock(
                        msg->reporter_id, msg->session_id,
                        msg->seq, msg->tx_hex);
                    break;
                case ips_msgs::msg::UwbAnchorReport::TYPE_SLAVE_MASTER:
                    engine_->add_slave_master_rx(
                        msg->reporter_id, msg->source_id,
                        msg->seq, msg->rx_hex);
                    break;
                case ips_msgs::msg::UwbAnchorReport::TYPE_SLAVE_TAG:
                    engine_->add_slave_tag_rx(
                        msg->reporter_id, msg->source_id,
                        msg->seq, msg->rx_hex);
                    break;
                default:
                    break;
            }
        } catch (const std::exception& e) {
            RCLCPP_WARN(get_logger(), "engine ingest error: %s", e.what());
        }
    }

    void on_event(const ips_msgs::msg::SessionEvent::SharedPtr msg) {
        std::lock_guard<std::mutex> lock(engine_mu_);
        if (msg->kind == ips_msgs::msg::SessionEvent::KIND_RESET ||
            msg->kind == ips_msgs::msg::SessionEvent::KIND_AUTO_DETECTED_RESTART) {
            engine_->handle_ma_reset(msg->device_id, msg->session_id);
        } else if (msg->kind == ips_msgs::msg::SessionEvent::KIND_HELLO) {
            engine_->handle_ma_hello(msg->device_id, msg->session_id);
        }
    }

    // ---- Engine callback: publish corrected ToA ----------------------
    // Runs INSIDE on_report under engine_mu_ — already serialized.
    void on_corrected_toa(const CorrectedToAResult& r) {
        ips_msgs::msg::CorrectedToA m;
        m.header.stamp    = now();
        m.header.frame_id = "uwb";
        m.tag_id          = r.tag_id;
        m.slave_id        = r.slave_id;
        m.tag_seq         = r.tag_seq;
        m.sync_seq        = r.sync_seq;
        m.toa_corrected_s = static_cast<double>(r.toa_final_dtu) * DTU_SECONDS;
        m.toa_li_s        = static_cast<double>(r.toa_li_dtu)    * DTU_SECONDS;
        m.toa_kf_s        = static_cast<double>(r.toa_kf_dtu)    * DTU_SECONDS;
        m.prop_delay_s    = static_cast<double>(r.prop_delay_dtu)* DTU_SECONDS;
        m.delta_k         = r.delta_k;
        m.theta_k_s       = static_cast<double>(r.theta_k_dtu)   * DTU_SECONDS;
        m.kf_used         = r.kf_used;
        pub_toa_->publish(m);
    }

    // ---- Status snapshot ---------------------------------------------
    void publish_status() {
        ips_msgs::msg::SyncStatus msg;
        {
            std::lock_guard<std::mutex> lock(engine_mu_);
            const auto status      = engine_->get_sync_status();
            msg.reset_count             = engine_->reset_count();
            msg.dropped_packets_on_reset = engine_->dropped_packets_on_reset();
            for (const auto& [sid, st] : status) {
                msg.slave_ids.push_back(static_cast<uint8_t>(sid));
                msg.sync_count.push_back(st.sync_count);
                msg.li_ready.push_back(st.li_ready);
                msg.kf_converged.push_back(st.kf_converged);
                msg.kf_updates.push_back(st.kf_updates);
                msg.delta_k.push_back(st.delta_k.value_or(0.0));
                msg.theta_k_s.push_back((st.theta_k_us.value_or(0.0)) * 1e-6);
                msg.kf_drift.push_back(st.kf_drift.value_or(0.0));
                msg.kf_drift_rate.push_back(st.kf_drift_rate.value_or(0.0));
            }
        }
        msg.header.stamp    = now();
        msg.header.frame_id = "uwb";
        pub_status_->publish(msg);
    }

    std::unique_ptr<SyncEngine> engine_;
    std::mutex                  engine_mu_;

    rclcpp::Subscription<ips_msgs::msg::UwbAnchorReport>::SharedPtr sub_reports_;
    rclcpp::Subscription<ips_msgs::msg::SessionEvent>::SharedPtr    sub_events_;
    rclcpp::Publisher<ips_msgs::msg::CorrectedToA>::SharedPtr       pub_toa_;
    rclcpp::Publisher<ips_msgs::msg::SyncStatus>::SharedPtr         pub_status_;
    rclcpp::TimerBase::SharedPtr                                    status_timer_;
};

} // namespace ips

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ips::ClockSyncNode>());
    rclcpp::shutdown();
    return 0;
}

File: software/ips_jazzy_ws/src/ips_nodes/ips_nodes/clock_sync_node.py
"""clock_sync_node — LI-KF clock synchronization (Zhang et al., 2024).

Subscribes to raw UWB anchor reports + session events, drives an
internal SyncEngine, and emits a CorrectedToA message for every blink
that completes synchronization (Step 2 LI + Step 3 KF + Step 4 prop delay).

Inputs:
    /uwb/anchor_reports  (ips_msgs/UwbAnchorReport, BEST_EFFORT)
    /uwb/session_events  (ips_msgs/SessionEvent,    RELIABLE)

Outputs:
    /uwb/corrected_toa   (ips_msgs/CorrectedToA, RELIABLE)
    /uwb/sync_status     (ips_msgs/SyncStatus,   RELIABLE)
"""

import threading

import rclpy
from rclpy.node import Node

from ips_msgs.msg import (
    UwbAnchorReport,
    SessionEvent,
    CorrectedToA,
    SyncStatus,
)

from ips_nodes.algorithms import sync_engine as se
from ips_nodes.common import (
    QOS_SENSOR_BEST_EFFORT_DEEP,
    QOS_STATE_RELIABLE,
    REPORT_TYPE_MASTER_CLOCK,
    REPORT_TYPE_SLAVE_MASTER,
    REPORT_TYPE_SLAVE_TAG,
    EVENT_RESET,
    EVENT_AUTO_RESTART,
    DTU_S,
)


class ClockSyncNode(Node):
    """Wraps SyncEngine in a ROS2 node with online callback emission."""

    def __init__(self) -> None:
        super().__init__('clock_sync')

        # ---- Parameters ------------------------------------------------
        self.declare_parameter('status_period_s', 1.0)
        self.declare_parameter('kf_enabled', True)
        # Optional CSV dump for offline analysis (default off in ROS2).
        self.declare_parameter('write_debug_csv', False)

        self._status_period: float = float(self.get_parameter('status_period_s').value)
        kf_enabled: bool = bool(self.get_parameter('kf_enabled').value)
        write_csv: bool = bool(self.get_parameter('write_debug_csv').value)

        # Apply KF flag to the algorithm module before instantiating engine.
        # se.KF_ENABLED = kf_enabled

        # ---- Engine ----------------------------------------------------
        # Lock guards engine + emission queue from concurrent access
        # (subscription callbacks are invoked from rclpy executor threads).
        self._engine_lock = threading.Lock()
        self._engine = se.SyncEngine(
            toa_callback=self._on_corrected_toa_record,
            write_csv=write_csv,
            kf_enabled=kf_enabled, # <--  BARIS Update 7 May
        )

        # ---- Subscriptions --------------------------------------------
        self._sub_reports = self.create_subscription(
            UwbAnchorReport,
            '/uwb/anchor_reports',
            self._on_report,
            QOS_SENSOR_BEST_EFFORT_DEEP,
        )
        self._sub_events = self.create_subscription(
            SessionEvent,
            '/uwb/session_events',
            self._on_event,
            QOS_STATE_RELIABLE,
        )

        # ---- Publishers -----------------------------------------------
        self._pub_toa = self.create_publisher(
            CorrectedToA, '/uwb/corrected_toa', QOS_STATE_RELIABLE
        )
        self._pub_status = self.create_publisher(
            SyncStatus, '/uwb/sync_status', QOS_STATE_RELIABLE
        )

        # ---- Status timer ---------------------------------------------
        self._status_timer = self.create_timer(
            self._status_period, self._publish_status
        )

        self.get_logger().info(
            f'clock_sync ready (KF={kf_enabled}, write_csv={write_csv})'
        )

    # ------------------------------------------------------------------
    # Subscription callbacks
    # ------------------------------------------------------------------
    def _on_report(self, msg: UwbAnchorReport) -> None:
        try:
            with self._engine_lock:
                if msg.report_type == REPORT_TYPE_MASTER_CLOCK:
                    self._engine.add_master_clock(
                        msg.reporter_id,
                        msg.session_id,
                        msg.seq,
                        msg.tx_hex,
                    )
                elif msg.report_type == REPORT_TYPE_SLAVE_MASTER:
                    self._engine.add_slave_master_rx(
                        msg.reporter_id,
                        msg.source_id,
                        msg.seq,
                        msg.tx_hex,
                        msg.rx_hex,
                    )
                elif msg.report_type == REPORT_TYPE_SLAVE_TAG:
                    self._engine.add_slave_tag_rx(
                        msg.reporter_id,
                        msg.source_id,
                        msg.seq,
                        msg.rx_hex,
                    )
        except Exception as exc:
            self.get_logger().warning(f'engine ingest error: {exc}')

    def _on_event(self, msg: SessionEvent) -> None:
        if msg.kind in (EVENT_RESET, EVENT_AUTO_RESTART):
            with self._engine_lock:
                self._engine.handle_ma_reset(msg.device_id, msg.session_id)
            self.get_logger().warning(
                f'engine reset triggered by event kind={msg.kind} '
                f'device={msg.device_id} session={msg.session_id}'
            )

    # ------------------------------------------------------------------
    # Engine callback (runs INSIDE _on_report under engine_lock)
    # ------------------------------------------------------------------
    def _on_corrected_toa_record(self, rec: dict) -> None:
        """Called by SyncEngine for every (slave, blink) pair that completes."""
        msg = CorrectedToA()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'uwb'
        msg.tag_id = int(rec['tag_id'])
        msg.slave_id = int(rec['slave_id'])
        msg.tag_seq = int(rec['tag_seq'])
        msg.sync_seq = int(rec['sync_seq'])
        msg.toa_corrected_s = float(rec['toa_final_dtu']) * DTU_S
        msg.toa_li_s = float(rec['toa_li_dtu']) * DTU_S
        msg.toa_kf_s = float(rec['toa_kf_dtu']) * DTU_S
        msg.prop_delay_s = float(rec['prop_delay_dtu']) * DTU_S
        msg.delta_k = float(rec['delta_k'])
        msg.theta_k_s = float(rec['theta_k_dtu']) * DTU_S
        msg.kf_used = bool(rec['kf_used'])
        self._pub_toa.publish(msg)

    # ------------------------------------------------------------------
    # Status snapshot
    # ------------------------------------------------------------------
    def _publish_status(self) -> None:
        with self._engine_lock:
            status = self._engine.get_sync_status()
            reset_count = self._engine.reset_count
            dropped = self._engine.dropped_packets_on_reset

        msg = SyncStatus()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = 'uwb'
        msg.reset_count = reset_count
        msg.dropped_packets_on_reset = dropped

        ids = sorted(status.keys())
        msg.slave_ids = ids
        msg.sync_count = [int(status[i]['sync_count']) for i in ids]
        msg.li_ready = [bool(status[i]['li_ready']) for i in ids]
        msg.kf_converged = [bool(status[i]['kf_converged']) for i in ids]
        msg.kf_updates = [int(status[i]['kf_updates']) for i in ids]
        msg.delta_k = [float(status[i]['delta_k'] or 0.0) for i in ids]
        # theta_k_us in engine status → convert to seconds for the message
        msg.theta_k_s = [
            float((status[i]['theta_k_us'] or 0.0) * 1e-6) for i in ids
        ]
        msg.kf_drift = [float(status[i]['kf_drift'] or 0.0) for i in ids]
        msg.kf_drift_rate = [
            float(status[i]['kf_drift_rate'] or 0.0) for i in ids
        ]
        self._pub_status.publish(msg)

    # ------------------------------------------------------------------
    def destroy_node(self) -> bool:
        try:
            with self._engine_lock:
                self._engine.close()
        except Exception:
            pass
        return super().destroy_node()


def main(args=None) -> None:
    rclpy.init(args=args)
    node = ClockSyncNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
